File: leetCode/LinkedList/reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let prev = null;
  let next = null;
  let current = head;
  while (current !== null) {
    next = current.next; // [2,3,4,5] // [3,4,5] // [4,5] // [5] // null
    current.next = prev; // [1,null]  // [2,1,null] // [3,2,1,null] // [4,3,2,1,null] // [5,4,3,2,1,null]
    prev = current; // [1,null]  // [2,1,null] // [3,2,1,null] // [4,3,2,4,null] // [5,4,3,2,1,null]
    current = next; // [2,3,4,5] // [3,4,5]    // [4,5] // [5] // null
  }
  return prev;
};

File: leetCode/LinkedList/reverse-linked-list.py
# 206. Reverse Linked List

# Given the head of a singly linked list, reverse the list, and return the reversed list.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummyNode = None

        while(head):
            node = dummyNode
            dummyNode = ListNode(head.val,node)
            head = head.next

        return dummyNode

# solution 2:
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, curr = None, head

        while(curr):
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        return prev

# solution 3: Recursion

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        newHead = head
        if head.next:
            newHead = self.reverseList(head.next)
            head.next.next = head

        head.next = None

        return newHead


